File: OS_kernel/py/06_communication.py
"""
============================================================
KERNEL AGENTIX OS — 06_communication.py
Module 6: Communication Interface
Transmisi Valensi Antar-Agen (Valence Bus)
============================================================
Depends on: corekernel.py
"""
from corekernel import AgentixKernel, Vector, Semantic


class CommunicationInterface(AgentixKernel):
    def __init__(self, config=None):
        super().__init__(config)
        self._bus      = []
        self._channels = {}
        self._inbox    = {}
        self._max_age  = (config or {}).get('comm', {}).get('max_age', 30)

    def broadcast(self, sender_id, channel, valence, pos=None, data=None):
        packet = {
            'sender_id': sender_id,
            'channel':   channel,
            'valence':   max(-1.0, min(1.0, valence)),
            'pos':       list(pos) if pos else None,
            'data':      dict(data or {}),
            'born':      self.iteration,
        }
        self._bus.append(packet)
        self._channels.setdefault(channel, []).append(packet)

    def send(self, sender_id, target_id, valence, data=None):
        self._inbox.setdefault(target_id, []).append({
            'sender_id': sender_id,
            'valence':   valence,
            'data':      dict(data or {}),
            'born':      self.iteration,
        })

    def read_inbox(self, agent_id):
        msgs = self._inbox.get(agent_id, [])
        self._inbox[agent_id] = []
        return msgs

    def scan_channel(self, channel, reader_pos=None, radius=float('inf')):
        result = []
        for p in self._channels.get(channel, []):
            age      = self.iteration - p['born']
            age_mask = int(age <= self._max_age)
            freshness = Semantic.decay(1.0, 0.97 ** age)
            spatial_w = 1.0
            if reader_pos and p['pos']:
                dist      = Vector.dist(reader_pos, p['pos'])
                spatial_w = Semantic.proximity(dist, radius)
            weight = age_mask * freshness * spatial_w
            if weight > 1e-4:
                result.append({**p, 'weight': weight})
        return result

    def aggregate_channel(self, channel, reader_pos=None, radius=float('inf')):
        packets = self.scan_channel(channel, reader_pos, radius)
        if not packets: return 0.0
        total_w = sum(p['weight'] for p in packets) + 1e-12
        sum_val = sum(p['valence'] * p['weight'] for p in packets)
        return sum_val / total_w

    def channel_consensus(self, channel):
        agg = self.aggregate_channel(channel)
        return (1 if agg > 0 else (-1 if agg < 0 else 0))

    def proximity_scan(self, reader_pos, radius):
        now = self.iteration
        out = []
        for p in self._bus:
            age     = now - p['born']
            alive   = age <= self._max_age
            in_range = not p['pos'] or Vector.dist(reader_pos, p['pos']) <= radius
            if alive and in_range:
                out.append({**p, 'dist': Vector.dist(reader_pos, p['pos']) if p['pos'] else 0.0})
        return out

    def flush_expired(self):
        now = self.iteration
        self._bus = [p for p in self._bus if (now - p['born']) <= self._max_age]
        for ch in list(self._channels):
            self._channels[ch] = [p for p in self._channels[ch] if (now - p['born']) <= self._max_age]

    def relay(self, packet, relayer_id, decay=0.8):
        self.broadcast(
            relayer_id,
            packet['channel'],
            packet['valence'] * decay,
            packet.get('pos'),
            {**packet.get('data', {}), 'relayed': True},
        )

File: OS_kernel/js/06_communication.js
/**
 * ============================================================
 * KERNEL AGENTIX OS — 06_communication.js
 * Module 6: Communication Interface
 * Transmisi Valensi Antar-Agen (Valence Bus)
 * ============================================================
 * Depends on: corekernel.js
 *
 * Model: Agents broadcast "valence packets" to a shared bus.
 * Recipients read the bus and integrate signals.
 * No direct agent-to-agent calls → pure relational broadcast.
 */
import { AgentixKernel, Vector, Semantic } from './corekernel.js';

class CommunicationInterface extends AgentixKernel {
    constructor(config) {
        super(config);
        this._bus       = [];       // global valence packet bus
        this._channels  = {};       // named channels { chName: [packets] }
        this._inbox     = {};       // per-agent inbox { agentId: [packets] }
        this._maxAge    = config?.comm?.maxAge || 30;  // ticks before packet expires
    }

    // ─────────────────────────────────────────
    // BROADCAST — agent emits a valence packet
    // packet: { senderId, channel, valence, pos, data, born }
    // ─────────────────────────────────────────
    broadcast(senderId, channel, valence, pos = null, data = {}) {
        const packet = {
            senderId,
            channel,
            valence: Math.max(-1, Math.min(1, valence)),
            pos:     pos ? [...pos] : null,
            data:    { ...data },
            born:    this.iteration,
        };
        this._bus.push(packet);

        if (!this._channels[channel]) this._channels[channel] = [];
        this._channels[channel].push(packet);
    }

    // ─────────────────────────────────────────
    // DIRECTED SEND — targeted agent inbox
    // ─────────────────────────────────────────
    send(senderId, targetId, valence, data = {}) {
        if (!this._inbox[targetId]) this._inbox[targetId] = [];
        this._inbox[targetId].push({
            senderId, valence, data, born: this.iteration
        });
    }

    // ─────────────────────────────────────────
    // READ INBOX — retrieve & clear messages for agent
    // ─────────────────────────────────────────
    readInbox(agentId) {
        const msgs = this._inbox[agentId] || [];
        this._inbox[agentId] = [];
        return msgs;
    }

    // ─────────────────────────────────────────
    // CHANNEL SCAN — read all packets on a channel
    // Filters by age (expired packets return 0 weight)
    // ─────────────────────────────────────────
    scanChannel(channel, readerPos = null, radius = Infinity) {
        return (this._channels[channel] || []).map(p => {
            const age      = this.iteration - p.born;
            const ageMask  = +(age <= this._maxAge);
            const freshness = Semantic.decay(1.0, 0.97 ** age);

            let spatialW = 1.0;
            if (readerPos && p.pos) {
                const dist = Vector.dist(readerPos, p.pos);
                spatialW   = Semantic.proximity(dist, radius);
            }

            return {
                ...p,
                weight: ageMask * freshness * spatialW,
            };
        }).filter(p => p.weight > 1e-4);
    }

    // ─────────────────────────────────────────
    // CHANNEL AGGREGATE — weighted sum of valences
    // ─────────────────────────────────────────
    aggregateChannel(channel, readerPos = null, radius = Infinity) {
        const packets = this.scanChannel(channel, readerPos, radius);
        if (packets.length === 0) return 0;
        const totalW  = packets.reduce((s, p) => s + p.weight, 0) + 1e-12;
        const sumVal  = packets.reduce((s, p) => s + p.valence * p.weight, 0);
        return sumVal / totalW;
    }

    // ─────────────────────────────────────────
    // CONSENSUS — majority valence direction on channel
    // Returns +1, -1, or 0
    // ─────────────────────────────────────────
    channelConsensus(channel) {
        const agg = this.aggregateChannel(channel);
        return Math.sign(agg);
    }

    // ─────────────────────────────────────────
    // PROXIMITY BROADCAST SCAN
    // Agent reads all nearby broadcasts regardless of channel
    // ─────────────────────────────────────────
    proximityScan(readerPos, radius) {
        return this._bus
            .filter(p => {
                const age     = this.iteration - p.born;
                const alive   = age <= this._maxAge;
                const inRange = !p.pos || Vector.dist(readerPos, p.pos) <= radius;
                return alive && inRange;
            })
            .map(p => ({
                ...p,
                dist: p.pos ? Vector.dist(readerPos, p.pos) : 0,
            }));
    }

    // ─────────────────────────────────────────
    // FLUSH — remove expired packets each tick
    // ─────────────────────────────────────────
    flushExpired() {
        const maxAge = this._maxAge;
        const now    = this.iteration;
        this._bus = this._bus.filter(p => (now - p.born) <= maxAge);
        Object.keys(this._channels).forEach(ch => {
            this._channels[ch] = this._channels[ch].filter(p => (now - p.born) <= maxAge);
        });
    }

    // ─────────────────────────────────────────
    // TOPOLOGY RELAY — relay packet to neighbors
    // Useful for mesh/chain propagation
    // ─────────────────────────────────────────
    relay(packet, relayerId, decay = 0.8) {
        this.broadcast(
            relayerId,
            packet.channel,
            packet.valence * decay,
            packet.pos,
            { ...packet.data, relayed: true }
        );
    }
}

export { CommunicationInterface };
